icu4j/main/classes/core/src/com/ibm/icu/text/MeasureFormat.java

```diff
@@ -861,116 +859,77 @@ private static DateFormat loadNumericDurationFormat(ICUResourceBundle r, String
+            pattern = numericFormatters.getHourMinuteSecond();
+            if (hms[1] == null)
+                hms[1] = 0;
+            hms[1] = Math.floor(hms[1].doubleValue());
+            hms[0] = Math.floor(hms[0].doubleValue());
+        } else if (hms[0] != null && hms[1] != null) { // "hm"
+            pattern = numericFormatters.getHourMinute();
+            hms[0] = Math.floor(hms[0].doubleValue());
+        } else if (hms[1] != null && hms[2] != null) { // "ms"
+            pattern = numericFormatters.getMinuteSecond();
+            hms[1] = Math.floor(hms[1].doubleValue());
+        } else { // h m s, handled outside formatNumeric. No value is also an error.
             throw new IllegalStateException();
         }
-    }
 
-    // Formats a duration as 5:00:37 or 23:59.
-    // duration is a particular duration after epoch.
-    // formatter is a hour-minute-second, hour-minute, or minute-second formatter.
-    // smallestField denotes what the smallest field is in duration: either
-    // hour, minute, or second.
-    // smallestAmount is the value of that smallest field. for 5:00:37.3,
-    // smallestAmount is 37.3. This smallest field is formatted with this object's
-    // NumberFormat instead of formatter.
-    // appendTo is where the formatted string is appended.
-    private void formatNumeric(
-            Date duration,
-            DateFormat formatter,
-            DateFormat.Field smallestField,
-            Number smallestAmount,
-            Appendable appendTo) {
-        // Format the smallest amount ahead of time.
-        String smallestAmountFormatted;
-
-        // Format the smallest amount using this object's number format, but keep track
-        // of the integer portion of this formatted amount. We have to replace just the
-        // integer part with the corresponding value from formatting the date. Otherwise
-        // when formatting 0 minutes 9 seconds, we may get "00:9" instead of "00:09"
-        FieldPosition intFieldPosition = new FieldPosition(NumberFormat.INTEGER_FIELD);
-        FormattedNumber result = getNumberFormatter().format(smallestAmount);
-        result.nextFieldPosition(intFieldPosition);
-        smallestAmountFormatted = result.toString();
-        // Give up if there is no integer field.
-        if (intFieldPosition.getBeginIndex() == 0 && intFieldPosition.getEndIndex() == 0) {
-            throw new IllegalStateException();
-        }
+        // We can create it on demand, but all of the patterns (right now) have mm and ss.
+        // So unless it is hours only we will need a 0-padded 2 digits formatter.
+        LocalizedNumberFormatter numberFormatter2 = numberFormatter.integerWidth(IntegerWidth.zeroFillTo(2));
+        FormattedStringBuilder fsb = new FormattedStringBuilder();
+
+        boolean protect = false;
+        for (int i = 0; i < pattern.length(); i++) {
+            Character c = pattern.charAt(i);
```

Comment:
Can you use type `char` not `Character` to avoid the boxing?

Reply:
I used Character so that I can do `c.toString()`
But I can do `char` and `String.valueOf(c)`.
Not sure if the performance makes a difference, but well, does not hurt :-)